float32 x_p
float32 y_p
float32 x_c
float32 y_c
float32 x_n
float32 y_n
float32 x_n_1
float32 y_n_1